File: ios/maimaid/Services/BackendConfig.swift
import Foundation

enum BackendConfig {
    static var baseURL: URL? {
        guard let value = AppInfo.configuredString(for: BundleInfoKeys.backendURL) else {
            return nil
        }
        return URL(string: value)
    }

    static var webAuthBaseURL: URL? {
        if
            let configured = AppInfo.configuredString(for: BundleInfoKeys.backendAuthURL),
            let configuredURL = URL(string: configured)
        {
            return configuredURL
        }

        guard let baseURL else {
            return nil
        }

        // Local default: backend runs on :8787, dashboard auth page runs on :3000.
        if baseURL.port == 8787, let host = baseURL.host {
            if host == "localhost" || host == "127.0.0.1" {
                var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)
                components?.port = 3000
                components?.query = nil
                components?.fragment = nil
                components?.path = "/"
                return components?.url
            }
        }

        return baseURL
    }
    
    static func endpoint(_ path: String) -> URL? {
        guard let baseURL else {
            return nil
        }

        let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return baseURL
        }

        let segments = trimmed.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: false)
        let pathSegment = String(segments[0])
        let pathURL = pathSegment.isEmpty ? baseURL : baseURL.appending(path: pathSegment)

        guard segments.count == 2 else {
            return pathURL
        }

        var components = URLComponents(url: pathURL, resolvingAgainstBaseURL: false)
        components?.percentEncodedQuery = String(segments[1])
        return components?.url
    }
}

File: android/app/src/main/kotlin/net/krtl/maimaid/core/data/remote/BackendConfig.kt
package net.krtl.maimaid.core.data.remote

import net.krtl.maimaid.BuildConfig
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

object BackendConfig {
    val baseUrl: HttpUrl?
        get() = BuildConfig.BACKEND_URL.trim().toHttpUrlOrNull()

    val webAuthBaseUrl: HttpUrl?
        get() {
            val configured = BuildConfig.BACKEND_AUTH_URL.trim()
            if (configured.isNotEmpty()) {
                return configured.toHttpUrlOrNull()
            }
            val base = baseUrl ?: return null
            if (base.port == 8787 && (base.host == "localhost" || base.host == "127.0.0.1")) {
                return base.newBuilder()
                    .port(3000)
                    .encodedPath("/")
                    .query(null)
                    .fragment(null)
                    .build()
            }
            return base
        }

    fun endpoint(path: String): HttpUrl? {
        val base = baseUrl ?: return null
        val trimmed = path.trim()
        if (trimmed.isEmpty()) return base
        return base.newBuilder().apply {
            val split = trimmed.split("?", limit = 2)
            val relativePath = split.first().trim('/')
            if (relativePath.isNotEmpty()) {
                addPathSegments(relativePath)
            }
            if (split.size == 2) {
                encodedQuery(split[1])
            }
        }.build()
    }
}

